tests/src/test/scala/org/apache/openwhisk/core/containerpool/test/ContainerPoolTests.scala

```diff
@@ -753,14 +757,102 @@ class ContainerPoolTests
-    containers(0).send(pool, ContainerRemoved)
-    containers(1).send(pool, ContainerRemoved)
+    containers(0).send(pool, ContainerRemoved(true))
+    containers(1).send(pool, ContainerRemoved(true))
     containers(2).expectMsg(Start(exec, memoryLimit))
     containers(3).expectMsg(Start(exec, memoryLimit))
     //make sure extra prewarms are not started
     containers(4).expectNoMessage(100.milliseconds)
     containers(5).expectNoMessage(100.milliseconds)
   }
+
+  it should "remove the prewarmed container after ttl time if unused" in {
+    val (containers, factory) = testContainers(2)
+    val feed = TestProbe()
+
+    val reactive: Option[ReactivePrewarmingConfig] = Some(ReactivePrewarmingConfig(0, 2, ttl, threshold, increment))
+    val pool =
+      system.actorOf(
+        ContainerPool
+          .props(
+            factory,
+            poolConfig(MemoryLimit.STD_MEMORY * 4),
+            feed.ref,
+            List(PrewarmingConfig(2, exec, memoryLimit, reactive))))
+    containers(0).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(1).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(0).send(pool, NeedWork(preWarmedData(exec.kind, expires = Some(ttl.fromNow))))
+    containers(1).send(pool, NeedWork(preWarmedData(exec.kind, expires = Some(ttl.fromNow))))
+    stream.reset()
+
+    // Make sure prewarmed containers can be deleted from prewarmedPool due to unused
+    Thread.sleep(3.seconds.toMillis)
+
+    pool ! AdjustPrewarmedContainer
+    containers(0).expectMsg(Remove)
+    containers(1).expectMsg(Remove)
+    containers(0).send(pool, ContainerRemoved(false))
+    containers(1).send(pool, ContainerRemoved(false))
+
+    // Make sure prewarmed containers are deleted
+    Thread.sleep(2.seconds.toMillis)
+
+    stream.toString should include("prewarmed container is deleted")
+    stream.reset()
+  }
+
+  it should "supplement prewarmed container when doesn't have enough container to handle activation" in {
+    val (containers, factory) = testContainers(6)
+    val feed = TestProbe()
+
+    val deadline: Option[Deadline] = Some(ttl.fromNow)
+    val reactive: Option[ReactivePrewarmingConfig] = Some(ReactivePrewarmingConfig(0, 2, ttl, threshold, increment))
+    val pool =
+      system.actorOf(
+        ContainerPool
+          .props(
+            factory,
+            poolConfig(MemoryLimit.STD_MEMORY * 4),
+            feed.ref,
+            List(PrewarmingConfig(2, exec, memoryLimit, reactive))))
+    containers(0).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(1).expectMsg(Start(exec, memoryLimit, Some(ttl)))
+    containers(0).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+    containers(1).send(pool, NeedWork(preWarmedData(exec.kind, expires = deadline)))
+    stream.reset()
+
+    // Make sure prewarmed containers can be deleted from prewarmedPool due to unused
+    Thread.sleep(3.seconds.toMillis)
+
+    pool ! AdjustPrewarmedContainer
```

Comment:
This is a good test, but I think it would also be good to a) move `ContainerPool.adjustPrewarmedContainer` to `object ContainerPool` and b) make additional tests dedicated to all logic in this function (starting count, current count, desired count, min count, max count, reactive config, nonreactive config, expired removal, counting cold starts)

Reply:
* Regarding move `class ContainerPool's adjustPrewarmedContainer ` to `object ContainerPool`
I tried, it seems it  is very complex, because adjustPrewarmedContainer  used several class variable, e.g. prewarmedPool, prewarmStartingPool, getReactiveCold(it uses coldStartCount), if do that, other places need to add these variables well when invoke `adjustPrewarmedContainer `
* Regarding `additional tests dedicated to all logic in this function`
I wrote extra test cases for that.

Reply:
Passing the map parameters is not a problem - see `ContainerPool.schedule`.
Function results can be changed from `Unit` to (Seq[ActorRef], Seq[(String,ByteSize,Int)]) representing (prewarms to delete, prewarms to create). My concern is that the sleep and message handling logic is very complicated, and may miss some edge cases in this critical function, so isolating it in a better testable way would give more confidence.

Reply:
hm...
Can you check this in you local env? 
i found that not only the problem of `passing the map parameter`, still has other problem(it seems not easy to describe clearly)， e.g. calls between variables and methods are too complicated

Reply:
Here is some code - not all to be moved to object, but the main logic to determine a) how many prewarms to create and b) which prewarms to remove:
```
  def adjustPrewarmedContainer(init: Boolean, scheduled: Boolean): Unit = {
    //fill in missing prewarms
    ContainerPool
      .increasePrewarms(init, scheduled, coldStartCount, prewarmConfig, prewarmedPool, prewarmStartingPool)
      .foreach { c =>
        (1 to c._2).foreach { _ =>
          val config = c._1
          prewarmContainer(config.exec, config.memoryLimit, config.reactive.map(_.ttl))
        }
      }
    if (scheduled) {
      //on scheduled time, remove expired prewarms
      ContainerPool.removeExpired(prewarmConfig, prewarmedPool).foreach { a =>
        a ! Remove
        prewarmedPool = prewarmedPool - a
      }
      //on scheduled time, emit cold start counter metric with memory + kind
      coldStartCount foreach { coldStart =>
        val coldStartKey = coldStart._1
        MetricEmitter.emitCounterMetric(
          LoggingMarkers.CONTAINER_POOL_PREWARM_COLDSTART(coldStartKey.memory.toString, coldStartKey.kind))
      }
      //   then clear coldStartCounts each time scheduled event is processed to reset counts
      coldStartCount = immutable.Map.empty[ColdStartKey, Int]
    }
  }
```
Then in object add:
```
  def removeExpired(prewarmConfig: List[PrewarmingConfig], prewarmedPool: Map[ActorRef, PreWarmedData])(
    implicit logging: Logging): List[ActorRef] = {
    prewarmConfig.flatMap { config =>
      val kind = config.exec.kind
      val memory = config.memoryLimit
      config.reactive
        .map { _ =>
          val expiredPrewarmedContainer = prewarmedPool
            .filter { warmInfo =>
              warmInfo match {
                case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if p.isExpired() => true
                case _                                                                  => false
              }
            }
          // emit expired container counter metric with memory + kind
          MetricEmitter.emitCounterMetric(LoggingMarkers.CONTAINER_POOL_PREWARM_EXPIRED(memory.toString, kind))
          logging.info(
            this,
            s"[kind: ${kind} memory: ${memory.toString}] removed ${expiredPrewarmedContainer.size} expired prewarmed container")
          expiredPrewarmedContainer.keys
        }
        .getOrElse(List.empty)
    }
  }
  def increasePrewarms(
    init: Boolean,
    scheduled: Boolean,
    coldStartCount: Map[ColdStartKey, Int],
    prewarmConfig: List[PrewarmingConfig],
    prewarmedPool: Map[ActorRef, PreWarmedData],
    prewarmStartingPool: Map[ActorRef, (String, ByteSize)])(implicit logging: Logging): Map[PrewarmingConfig, Int] = {
    prewarmConfig.map { config =>
      val kind = config.exec.kind
      val memory = config.memoryLimit

      val runningCount = prewarmedPool.count {
        // done starting, and not expired
        case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if !p.isExpired() => true
        // started but not finished starting (or expired)
        case _ => false
      }
      val startingCount = prewarmStartingPool.count(p => p._2._1 == kind && p._2._2 == memory)
      val currentCount = runningCount + startingCount

      logging.info(
        this,
        s"[kind: ${kind} memory: ${memory.toString}] currentCount: ${currentCount} prewarmed container")

      // determine how many are needed
      val desiredCount: Int =
        if (init) config.initialCount
        else {
          if (scheduled) {
            // scheduled/reactive config backfill
            config.reactive
              .map(c => getReactiveCold(coldStartCount, c, kind, memory).getOrElse(c.minCount)) //reactive -> desired is either cold start driven, or minCount
              .getOrElse(config.initialCount) //not reactive -> desired is always initial count
          } else {
            // normal backfill after removal - make sure at least minCount or initialCount is started
            config.reactive.map(_.minCount).getOrElse(config.initialCount)
          }
        }

      logging.info(
        this,
        s"found ${currentCount} started and ${startingCount} starting; ${if (init) "initing" else "backfilling"} ${desiredCount - currentCount} pre-warms to desired count: ${desiredCount} for kind:${config.exec.kind} mem:${config.memoryLimit.toString}")(
        TransactionId.invokerWarmup)
      (config, desiredCount)
    }.toMap
  }
  def getReactiveCold(coldStartCount: Map[ColdStartKey, Int],
                      config: ReactivePrewarmingConfig,
                      kind: String,
                      memory: ByteSize): Option[Int] = {
    coldStartCount.get(ColdStartKey(kind, memory)).map { value =>
      // Let's assume that threshold is `2`, increment is `1` in runtimes.json
      // if cold start number in previous minute is `2`, requireCount is `2/2 * 1 = 1`
      // if cold start number in previous minute is `4`, requireCount is `4/2 * 1 = 2`
      math.min(math.max(config.minCount, (value / config.threshold) * config.increment), config.maxCount)
    }
  }
```
WDYT?

Reply:
I fixed 2 bugs in your codes.
It is very good!